zebra-chain/src/transaction/unmined.rs

```diff
@@ -0,0 +1,171 @@
+//! Unmined Zcash transaction identifiers and transactions.
+//!
+//! Transaction versions 1-4 are uniquely identified by narrow transaction IDs,
+//! whether they have been mined or not,
+//! so Zebra and the Zcash network protocol don't use wide transaction IDs for them.
+//!
+//! Zebra's [`UnminedTxId`] and [`UnminedTx`] enums provide the correct unique ID for
+//! unmined transactions. They can be used to handle transactions regardless of version,
+//! and get the [`WtxId`] or [`Hash`] when required.
+
+use std::sync::Arc;
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest_derive::Arbitrary;
+
+use super::{
+    AuthDigest, Hash,
+    Transaction::{self, *},
+    WtxId,
+};
+
+use UnminedTxId::*;
+
+/// A unique identifier for an unmined transaction, regardless of version.
+///
+/// "The transaction ID of a version 4 or earlier transaction is the SHA-256d hash
+/// of the transaction encoding in the pre-v5 format described above.
+///
+/// The transaction ID of a version 5 transaction is as defined in [ZIP-244].
+///
+/// A v5 transaction also has a wtxid (used for example in the peer-to-peer protocol)
+/// as defined in [ZIP-239]."
+/// [Spec: Transaction Identifiers]
+///
+/// [ZIP-239]: https://zips.z.cash/zip-0239
+/// [ZIP-244]: https://zips.z.cash/zip-0244
+/// [Spec: Transaction Identifiers]: https://zips.z.cash/protocol/protocol.pdf#txnidentifiers
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+pub enum UnminedTxId {
+    /// A narrow unmined transaction identifier.
+    ///
+    /// Used to uniquely identify unmined version 1-4 transactions.
+    /// (After v1-4 transactions are mined, they can be uniquely identified using the same [`transaction::Hash`].)
+    Narrow(Hash),
+
+    /// A wide unmined transaction identifier.
+    ///
+    /// Used to uniquely identify unmined version 5 transactions.
+    /// (After v5 transactions are mined, they can be uniquely identified using only their `WtxId.id`.)
+    ///
+    /// For more details, see [`WtxId`].
+    Wide(WtxId),
+}
+
+impl From<Transaction> for UnminedTxId {
+    fn from(transaction: Transaction) -> Self {
+        // use the ref implementation, to avoid cloning the transaction
+        UnminedTxId::from(&transaction)
+    }
+}
+
+impl From<&Transaction> for UnminedTxId {
+    fn from(transaction: &Transaction) -> Self {
+        match transaction {
+            V1 { .. } | V2 { .. } | V3 { .. } | V4 { .. } => Narrow(transaction.into()),
+            V5 { .. } => Wide(transaction.into()),
+        }
+    }
+}
+
+impl From<WtxId> for UnminedTxId {
+    fn from(wtx_id: WtxId) -> Self {
+        Wide(wtx_id)
+    }
+}
+
+impl From<&WtxId> for UnminedTxId {
+    fn from(wtx_id: &WtxId) -> Self {
+        (*wtx_id).into()
+    }
+}
+
+impl UnminedTxId {
+    /// Return a new `UnminedTxId` using a v1-v4 legacy transaction ID.
```

Comment:
Nit: I'm wondering if it would make sense to remove `Return` from the documentation of each method. For the constructor, it could be `Create`, while for the getters it could simply be removed. I'm not sure if that actually makes things clearer :thinking:

Reply:
I try to follow whatever the Rust standard library does here.
So you're right, `Create` is better here.

Here's a `std` example:
https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.new

```suggestion
    /// Create a new `UnminedTxId` using a v1-v4 legacy transaction ID.
```